### DISCUSSION TITLE
Security implications of storing an unencrypted private key "rsa_key.pem"

### DISCUSSION THREAD

**Discussion by DonEstefan** [dani-garcia/vaultwarden#2796] [Q&A] [answered]
I noticed that the /data/ folder contains an unencrypted rsa private key "rsa_key.pem".

> [These files](https://github.com/dani-garcia/vaultwarden/wiki/Backing-up-your-vault#the-rsa_key-files) are used to sign the JWTs (authentication tokens) of users currently logged in.

Does it mean that the private key should be considered sensitive data, just like the [config.json](https://github.com/dani-garcia/vaultwarden/wiki/Backing-up-your-vault#the-configjson-file)?
Could a stolen rsa_key (e.g. from a compromised cloud backup) be abused to manipulate login sessions? Unfortunately the statement in the linked article leaves room for interpretation.

**Comment by stefan0xC** [comment]
You could probably also gain access to the admin panel with the private key somehow (if it's enabled and not further protected) but aside from that nothing more substantial should be possible e.g. there's no way to get access to a user's vault entries or sends without their vault password, and you can also not add yourself to an organization, confirming an emergency invite or something like that.

**Comment by jjlin** [comment, marked as answer]
I've added additional details to https://github.com/dani-garcia/vaultwarden/wiki/Backing-up-your-vault#the-rsa_key-files.

I think the most interesting attack possible involves first obtaining a copy of the database and using that to offline brute-force the master password of one or more users. This in itself would provide access to the secrets in that copy of the database. However, the database also
contains the unencrypted UUIDs needed to forge a login session to the server in order to obtain ongoing access to updated entries. This isn't likely to succeed unless you or your users have weak or reused master passwords, though.

On balance, I wouldn't consider it strictly required to encrypt the private key file unless you have particular concerns about your storage provider, but it's not that hard to encrypt it when you're already encrypting other things anyway, so you might as well. Personally, I [encrypt the entire backup](https://github.com/jjlin/vaultwarden-backup), including the database file.

**Reply by DonEstefan** [reply]
Thanks for the explanation and the documentation changes. It is much appreciated!
